### ISSUE TITLE
Migrate MC truth-matching code here

### ISSUE THREAD

**Issue by yipengsun** [umd-lhcb/lhcb-ntuples-gen#78]
Currently our MC truth-matching code is scattered around `plot_scripts`. We should migrate all of them here.

**Comment by afernez** [comment]
@yipengsun I have two questions for this:
- The first is just about setup of the code for [`truth_match.h`](https://github.com/umd-lhcb/lhcb-ntuples-gen/blob/044f0f1b38ce5741c5f331cfa19e9d24be90aafd/include/functor/rdx/truth_match.h), since I don't know how this code will be utilized within babymaker. I've committed an outline for how I'm planning to write the code; does this seem usable to you, or does it need to be organized differently (primarily w.r.t. how I can invoke the needed variables from the ntuples)?
- The second I think will require more work. The definitions of the flags are different between the `D*` and `D0` samples. The [`flag.h`](https://github.com/umd-lhcb/lhcb-ntuples-gen/blob/fe630bc47be26220e1931f91cc72a5acb9c0e7c1/include/functor/rdx/flag.h) that you wrote is based on Phoebe's flag definitions in [`AddB.C`](https://github.com/umd-lhcb/RDRDstRun1AnalysisPreservation/blob/master/proc/AddB.C), and I believe this script defines the flags for the `D*` sample only. I think the flags for the `D0` sample are defined in her [`AddD0B_temp.C`](https://github.com/umd-lhcb/RDRDstRun1AnalysisPreservation/blob/f78365bfc3000ef70ced2ed271ec42aa866c44af/proc/AddD0B_temp.C). Of course, both flag definitions are needed for the two sets of truth-matching selections. I could create a second file, `flag_D0.h`, that mimics the structure of your `flag.h` but has the flag definitions for the `D0` sample (also, I'd probably rename `flag.h` to `flag_Dst.h`). This will take more time than just copying over the truth-matching selections, though, since this part of Phoebe's code is more spreadout and the similar task of writing `flag.h` proved difficult to be careful with in the past.

**Comment by yipengsun** [comment]
I took a look at your skeleton, and it makes sense. However, given that now we **KNOW** the MC decay IDs (provided as the first argument), the truth-matching should be trivial (aside from the `keepme` flag), and we don't need Phoebe's flags at all, IMO.

**Comment by yipengsun** [comment]
With all the branches from `TupleToolSLTruth`, maybe we can have an easier way to truth-match our MC samples:

- Require the reconstructed B meson is a true B
- Require the `_True_Q2 > 100*100`, as that is the true momentum transfer directly from the B decay
- If the decay ID indicates a Tau Nu decay, require `_True_IsTauDecay`
- Based on sig, norm, D**, DDX, require that the direct B daughters have matching D mesons

**Comment by afernez** [comment]
To give an update where I am now, and also to comment on the selections present in Phoebe's truth-matching in general that aren't included in your bullets above:
I have most of the truth-matching code ("copy" of Phoebe's, but without the flags, and _hopefully_ at least a little clearer overall structure) written, but still have some work to do to finish the `D**` and `DD` truth-matching functions for the `D0` sample, as well as checking over the code/testing it (for any typos, mostly).
Selections I've noted she has in addition to what you propose above (for reference, if nothing else):
- She checks that B is correctly the mother/grandmother/etc of mu
- For the `D*` sample, she uses background categorization variables like `dst_bkgcat`, `d0_bkgcat`, and `b_bkgcat` as part of the truth-matching (recall variables described [here](https://twiki.cern.ch/twiki/pub/LHCb/Particle2MC/BkgCat_211013.key.pdf))
- For the `D0` sample, she has a cut on the `D0` mass (within 23.4MeV of PDG value)
- For the `D*` sample `D**` decays, she cuts out decays with two pions (in the D**->D* decay)

Checking that the mu is correctly a B daughter/granddaughter/etc is an obvious thing to be added, I think. Also, notably, because there's so many possible decays included in the decfiles for the DD decays, the truth-matching for the mu and D0/D* ends up being nontrivial there. I think this can be accounted for, one just has to be careful. For the other things, I think they can be easily implemented, I just wanted to list them to have everything in one place in plain English.

**Comment by afernez** [comment]
To note here: we had decided that it didn't really make sense to implement a full copy of Phoebe's truth-matching using the `TupleToolSLTruth` variables, since Phoebe's truth-matching ends up utilizing more information than can be extracted from just the `b0_TrueHadron_D(0,1,2)[_GD(0,1,2)]` variables. Instead, I'll implement a copy of Phoebe's truth-matching AND a truth-matching using these variables- in the end, I'll implement the latter truth-matching just as Yipeng [described above](https://github.com/umd-lhcb/lhcb-ntuples-gen/issues/78#issuecomment-890609131). I'll then compare the two truth-matchings on some subsets of `D*/D0 sig, norm, D**[_H,s], DDX` MC and we can decide which truth-matching procedure we want to proceed with. Perhaps of note, I'm thinking for all decay modes except `DDX`, the truth-matching using the `TupleToolSLTruth` variables will essentially be a subset of Phoebe's truth-matching. But for `DDX`, since we can actually distinguish more than one charm particle at each generation after the `B` now, rather than Phoebe's seemingly "loose" matching (summarized in my slides [here](https://github.com/umd-lhcb/group-talks/blob/d332f5437abe826191cc8615508d4899f5d40fcd/rdx/21-03-09_alex_run2_fullsim_validation.pdf)), we can actually ensure that the multiple charm mesons are present (basing the selections on the relevant dec files; this will be rather complicated, because of all the possible `B` decays); it'll be interesting to see if these selections for `DDX` end up being comparable to Phoebe's.

Also, @yipengsun I wanted to check my understanding of how `TupleToolSLTruth` numbers the hadronic daughters/granddaughters, which you had described a bit [here](https://github.com/umd-lhcb/lhcb-ntuples-gen/issues/77#issuecomment-885334372) and later in that issue. It looks to me that, for the particular decay mode, the hadrons of interest should be listed in "ascending" order (wrt relevance). For example, when I want to ensure `B0->D*`, I think I should be selecting `b0_TrueHadron_D0_ID==D*` (and ignore `D1`, `D2`), right? And to truth-match `B0->D0 Xc X` I'd select `b0_TrueHadron_D(0,1)_ID` to be charm mesons (ignoring `D2`)?

**Comment by yipengsun** [comment]
From the code snippet there, I never saw an explicit **"ascending"** ordering. However, what is probable is that indeed there's an ordering in the way these tracks get stored in the `.dst` file, so your way of checking should work (i.e. check `_D0` when we are working on `D**, D*, D` modes and check `_D0, _D1` when on `DD` modes). 

Still, to be save, I was checking all 3 daughters and make sure one of them is a `D**, D*, D`. But in principle your idea should work so let's start with yours and see if it is indeed working. If so, simpler's better. We just need to add a one-line comment on the ordering so we won't be confused when we look back 3 years later.

**Comment by yipengsun** [comment]
I also have a `is_D_meson` function implemented here: https://github.com/umd-lhcb/hammer-reweight/blob/70eeba9bef04b396462519a4590657a4cfc28489/include/utils.h#L78-L82.

**Comment by yipengsun** [comment]
In my reweighting script, I also require for all  **true** particles extracted by the `TupleToolSLTruth`, it has a invariant mass that is greater or equal to 0. This gives me about 75% of total truth-matching rate for `B0 -> D* Tau Nu`. Maybe this is what is needed?

**Comment by yipengsun** [comment]
@afernez I'm looking at your truth-matching code, and it seems to me that it will only tell if an event is truth-matched, without the ability to separate various `D**` components? Am I right? If so, perhaps I should add a branch that stores the true ID of up to 2 D mesons so we can separate the cocktail MCs for `D**` and `DD`?

**Comment by afernez** [comment]
Ah yeah I meant to make a comment here to describe the usage a bit. But the answer to your question is: it'll allow you to separate the `D**` components, but not `DD` (yet- I would have to add some code and use the daughter variables from `TupleToolSLTruth` to do this); it'll also perform the truth-matching for these separated `D**` cocktails (I'm not sure if you want this or not for what you're doing). For example, if you're looking specifically for `B0 -> D1 mu nu`, it'll perform the truth-matching (as defined by Phoebe) for the `B0 -> D**+ mu nu` decay and require the identified `D**+` to be `D1`.

For usage: there are two functions for the user to call for truth-matching, `MC_TRUTH_MATCH_DST` and `MC_TRUTH_MATCH_D0`, naturally for truth-matching for the `D*` and `D0` samples. The user provides, as arguments to these functions,
1. The decay id they want to truth match against
2. A bunch of truth information to do the truth-matching
3. Optionally a cocktail ID (only useful for `D**`, including heavy and strange `D**`) that they want to separate (like you do now). The idea is that the user provides the MC ID of the `D**` to be selected (eg. for `D1+` set `cocktail_id=10413`, or using the variables I defined in `include/pdg.h`, they could set `cocktail_id=PDG_ID_D1`). Notably, you have to choose the MC ID with the correct sign (ie. `D1+` and not `D1_0` for the example); the code should complain if a MC ID that isn't an actual cocktail for the decay is selected.
4. Optionally a boolean that designates (if truth-matching a `D**` decay, not including heavy or strange `D**`) if they want the decay to proceed with two pions or not (ie. like `D** -> D* pi pi` or only one pion, etc)

That should be all the directions that are needed, I think. Those user functions are located at the bottom of the truth-matching code, by the way. Do you think it would be useful for me to add a small subsection on the [step 2 ntuples with babymaker](https://umd-lhcb.github.io/lhcb-ntuples-gen/ntupling/step2_babymaker/) wiki page with what I wrote here?

**Comment by yipengsun** [comment]
I see what you are doing here. Just to check if my understanding is correct: If we want to separate the `D**` cocktail, we need to provide the desired `D**` PDG MC ID so that the truth-matching boolean will be true ONLY for this specific `D**` particle.

I was thinking along the lines of: For truth-matching, ALL `D**` will be considered OK, and we'll have separate booleans for each `D**` like `is_d1`, `is_d1st`.

Maybe I should provide two branches like `d_meson1_true_id` and `d_meson2_true_id` so that we can separate these cocktails more easily? Also these would be useful to separate the `DD` MC cocktails.

**Comment by yipengsun** [comment]
Continue with my previous post, with a bit more details:

Your truth-matching boolean will be generic, in the sense that for `D0` and `D*`, you match the exact decay modes; for `D**`, you match ALL valid `D**` candidates.

In addition, we use the `d_meson{1,2}_true_id` to separate the `D**` and `DD` cocktails with separate functions, like:

```cpp
Bool_t IS_D1(Int_t d_meson1_id, Int_t d_meson2_id) {
  if (d_meson2_id != 0) return false;
  if (TMath::Abs(d_meson1_id) == PDG_ID_D1) return true;
  return false;
}
```

**Comment by yipengsun** [comment]
In general, we should have these flags in our step-2 MC ntuples:
```shell
is_truth_matched  # this depends on D* or D0 tree, also depends on MC ID, like 11574011
is_d1
is_d1st
# additional D** flags
# additional DD flags
```

**Comment by afernez** [comment]
So just to summarize first- in our meeting Tuesday we decided it made the most sense to only have one branch in our postprocessed ntuples for the truth-matching. Instead of a boolean, we'll use an int that encodes both whether the truth-matching was successful and what kind of decay was matched to. As a first proposal for this, we could try a 7 digit int `truthmatch=d1d2d3d4d5d6d7` where

  1. `d1` = **1** if truth-matched for the `D*` sample, **2** if truth-matched for the `D0` sample. This should be redundant, since the filename of the postprocessed ntuple should specify this too, but I prefer the redundancy.
  2. `d2` = **0** if normalization(-like) decay (ie. with no tau), **1** if signal(-like) decay (ie. with a tau)
  3. `d3` = **0** if `B` meson is `B0==Bd`, **1** if `B==B-==Bu`, **2** if `B==Bs`
  4. `d4` = **0** if normalization/signal decay, **1** if (light) `D**`, **2** if (strange) `D**s`, **3** if (heavy) `D**H`, **4** if `DD`
  5. `d5` =
      1. if `d4==0`: **0**
      2. if `d4==1` or `d4==2`: **0** if `D**==D0*`, **1** if `D**==D1`, **2** `D**==D1'`, **3** if `D**==D2*` (note the `D**` sign can be deduced from `d3`, and strange quark presence can be deduced form `d4`)
      3. if `d4==3`: **0** if `D**H==D*(2S)`, **1** if `D**H==D(2S)`, **2** if `D**H==D(2750)`, **3** `D**H==D(3000)`
      4. if `d4==4`: **0** (for now)
  6. `d6` =
      1. if `d4!=3` or `d1==1`: **0**
      2. if `d4==3` and `d1==2`: **0** if no `D*`, **1** if `D*` is present (ie. `D**H->D*->D0`)
  7. `d7` =
      1. if `d4!=1`: **0**
      2. if `d4==1`: **0** if two pi decays (`D**->D(*,0) pi pi`) are cut out, **1** if the `D**` decays to two pi

And, if the truth-matching failed for the event, just have `truthmatch=-1`. This integer should then contain all the relevant info from the decay ID being matched against AND additional info separating cocktails AND additional info about cutting out two pion decays for `D**`.

When adding code to separate out the different `DD` decays, we may have to add another digit (beyond `d5`) because there are many possible decays.

Finally, for calculating `truthmatch`, one annoying thing that I would like to avoid, if it's possible, is having to manually reset the decay ID each time you run on a different MC ntuple. Right now, the user is responsible for setting the decay ID argument in the yml file. I *think* it should always be true that if you truth-match using the wrong decay ID, the truth-matching will fail and so will set `truthmatch=-1`, but I would want to test this out to be sure. If this is true, I think I can just avoid the user having to input the decay ID by just trying to truth-match against all possible decay IDs in my code. This will be slower, though.

**Comment by yipengsun** [comment]
Can we have something more symmetrical, like this: `-EDCBA` (a signed 5-digit number): `(D_meson)^C_B`

- The minus sign will be consistent w/ the D meson true ID
- If the D meson is charged, `A = 1`, otherwise `0`
- If the D meson has some subscrip, use `B` to denote
    - say, for `D` and `D*`, `B = 0`, 
    - for `D_0`, `B = 3`
    - for `D_1 = 1`
- If the D meson has a `*` superscript, `C` is 1, otherwise 0
- If the D meson contains a `s` quark, `D = 1`
- `E` is reserved for some special meaning, such as if the 2 `pi` decays are cut out

A failed truth-match will have a value of `0`, so we can just cut `truth_matched != 0` to remove events with failed truth-matching

Another branch like `d_extra` will record the meson for _DD_ MC samples. This can be purely based on MC true ID, and the `E` digit will be always 0.

**Comment by yipengsun** [comment]
After discussion, I have the following proposal:

There will be two integers from the truth-matching: `truthmatch` `d_extra_info` (names are chosen by me)

## `truthmatch`

It will be a 5-digit number: `-dcbaa`:

- The minus sign will be consistent w/ the _D_ meson true ID
- `aa` indicates the _D_ meson specie. Note that we treat `D0` and `D+` as the same specie in this case. More specifically:
    - `1`: `D`. **Note**: We can't start from `0` otherwise it may be confused with candidates that fail truth-matching
    - `2`: `D*`
    - `3`: `D*_0`
    - `4`: `D_1`
    - `5`: `D'_1`
    - `6`: `D*_2`
    - `7-99`: Additional _D_ meson variants (e.g. `D_s`)
    - **Note**: I reserved 2-digit for possible _D_ species just in case. 
 - `b`: Indicate if the _D_ meson is charged (`b = 1`) or not (`b = 0`)
 - `c`: `c = 0` for _Mu_, `c = 1` for _Tau_
 - `d`: Special meanings:
     - `d = 1`: Is _D**_ sample and the 2 pi NOT cut off
     - `d = 2`: is _DD_ sample

For `B -> D0 Mu Nu` decay, the number would be `00001`; for `B -> D* Tau Nu`, `01102`.

## `d_extra_info`

Save the specie-info of the other _D_ meson for the _DD_ sample. For all other samples, `d_extra_info = 0`. 
The scheme will be `-baa` and `aa` and `b` follows the same definition as above.

@manuelfs @afernez Please read my proposal and green-light it if you think it's reasonable.

One thing to note is that I used 2 digits `aa` for _D_ species. If we are super sure about that the number of _D_ species will be smaller than 9, then we can just use `a` instead of `aa`.

**Comment by yipengsun** [comment]
I updated my proposal above to make the common modes only 3-digit long.

**Comment by afernez** [comment]
Sure. So, building off your most recent proposal and taking our discussion in Tuesday's meeting into account-
For now, we can use a 4-digit `a1b1b2c1` scheme for indicating what decay an event was truth-matched to.
I think in the end there wasn't a definite preference between including `DD` cocktail info in extra preceding digits or in another int, and in any case, I haven't implemented anything in our truth-matching to distinguish all the various `DD` decays yet. For now, it just (as Phoebe does for her templates) distinguishes between normalization-like and signal-like DD decays and between decays from `B0` vs `B+`. We can encode this info in `a1b1b2c1` for the DD decays now and then when I implement code to separate the DD cocktails, we can figure out how to encode the additional info and whether to append to `truthmatch` or have a separate int.

Now, for `truthmatch=a1b1b2c1`, how about-
1. `truthmatch=0` will represent failed truth-matching
2. We won't use a sign to distinguish CC pairs for the "primary" `D` meson. We could add this in later, but for now, we'll have `truthmatch>=0`
3. `c1`=
    1. 0 if normalization(-like) (ie. without tau)
    2. 1 if signal(-like)
4. `b1b2` are two digits referring to the "primary" `D` meson. For non-DD decays, this is just the `D` that the `B` decays to; for DD decays, I'd propose having `b1b2=00`, and we can differentiate the `B0` vs `B+` using `a1`. And, for the non-DD decays, we can use the numbering scheme `b1b2=`
    1. `01` for `D0`, `02` for `D+`, `03` for `D*0`, `04` for `D*+`
    2. `10` for all (light) `D**^0` (it's required that the `D**` be possible for the decay, ie. that it's one of the `D**` listed in the respective dec file), `11` for `D0*^0`, `12` for `D1^0`, `13` for `D1'^0`, `14` for `D2*^0`
    3. `20` for all (light) `D**^+`, `21` for `D0*^+`, `22` for `D1^+`, `23` for `D1'^+`, `24` for `D2*^+`
    4. `30` for all heavy `D**H^0` (again, it's internally required that the `D**H` is possible), and (though Phoebe never separates these cocktails, I implemented the ability to do so, in case we're ever curious) `31` for `D*(2S)^+`, `42` for `D(2S)^+`, `43` for `D(2750)^+`, `44` for `D(3000)^+`
    5. `40` for all heavy `D**H^+`, and `41` for `D*(2S)^+`, `42` for `D(2S)^+`, `43` for `D(2750)^+`, `44` for `D(3000)^+`
    6. `50` for all strange `D**s` (again, internally required to be possible), and (again, not implemented by Phoebe, by just as optional functionality) `53` for `D1'_s`, `54` for `D2*_s` (to keep consistent with `D**` scheme; there aren't any "`D0*_s, D1_s`")
5. `a1`=
    1. `0` if not a `D**H` or DD decay, and `D**->D(*)pipi` decays are cut out if a (light) `D**` decay
    2. `1` if a `DD` decay from a `B0`
    3. `2` if a `DD` decay from a `B+`  
    4. `3` if a `D**H` decay where `D**H->D*->D` (useful because Phoebe separates this topology from `D**H->D0` directly in her templates for the `D0` sample)
    5. `4` if a `D**H` decay where `D**H->D`
    6. `5` if a (light) `D**` decay explicitly with two pion decays (`D**->D(*)pipi`) kept  
6. Just as a note: I'm making an underlying assumption that we can distinguish whether the event was reconstructed for `D*` or `D0` using other info from the ntuple (or the filename). This is needed to identify what truth-matching was done (eg. there is truth-matching for `B0->D**munu` for both `D*` and `D0`, but it's different for the two samples), and nothing I've said here encodes this info in `a1b1b2c1`

**Comment by yipengsun** [comment]
Wait, I thought the final scheme is like this: `a1b1b2b3b4c1`, and `b1b2` will be the absolute value of the true ID of the second D meson, and for 1 D meson case, `0`, a couple of examples:

- `B0->D*munu`:  `000040`
- `B-->D*0munu`: `000030`
- `B0->D0taunu`: `000011`
- `B0->D1'taunu`: 000231`
- `B0->D**H(->Dpipi)munu`: `4??400`
- `B+->D0X(->munuX)X`: `2??010`

Note that for `b1b2` and `b3b4`, they follow your particle map (e.g. `01 -> D0`, `23 -> D1'`). Or is it that we decide to NOT be explicit about the second D (or generic H) meson in this branch?

**Comment by afernez** [comment]
My memory probably wasn't perfect- I thought we had first agreed on a `d1d2a1b1b2c1` scheme with `d1d2` used for separating `DD` cocktails, but then someone said perhaps a second int may potentially be easier, so I considered it undecided and thought we could just choose in the future (when I implement this functionality in the code). In any case, now, I guess the pro/con list for these options (and the option with `a1` to the left) is
|Good Feature                                                             |`d1d2a1b1b2c1`     |`a1d1d2b1b2c1`      |`a1b1b2c1` *and* `d1d2`|
|---------------------------------------------------------|---------------------|----------------------|--------------------------|
|Back. comp. if `a2` added                                           |                               |:heavy_check_mark:|:heavy_check_mark:       |
|If not separating `DD`,<br>`truthmatch` is 4 digits\*|:heavy_check_mark:|                                |:heavy_check_mark:       |
|Only 1 coded int needed                                           |:heavy_check_mark:|:heavy_check_mark:|                                       |

*\*`truthmatch` would only become 6 digits for `DD`, `D**H`,`D**->D(*)pipi` (and possibly other special) decays*

Personally, I'd weight backward compatibility the lowest among these good features, so **my preference would be `d1d2a1b1b2c1`**.


For the `D**` numbering- I agree with the reasoning that having `b1b2=10,20` specifically to represent all `D**0/D**+` cocktails isn't really necessary as far as cuts are concerned (ie. `10<b1b2 && b1b2<20` is almost equally as easy as `b1b2==10`, and would rarely be used anyway). My reasoning was more that it helps with the symmetry of the numbering when considering `D**H` and `D**s`. In my proposed scheme, we really would only be generally using `b1b2=30,40,50` for `D**H, D**s`, but I thought maybe for debugging it could be useful to have the cocktail-separating ability implemented, and then the most natural way to separate *all* from *specific* cocktails was to have `b2==0` for *all* and `b2>0` for *specific*. It seemed unnatural to me to have this convention for `D**H` and `D**s` but then change it for `D**`.
By the way, I also chose to separate `D**` charges using the first digit `b1` (so that neutral `D**` are odd in `b1`, charged are even) to try to partially recover Yipeng's original idea to have the parity of `truthmatch` indicate the charge of the `D` meson. Do you think this makes the things more complex without much added value, or do you agree with the choice?

**Comment by yipengsun** [comment]
I'd say that `a1` is rarely used, because so far the only usages are for `DD` and `D**H` samples, and for `DD` samples, the `d1d2` will be populated anyway (we can also use `d1d2` to record the true ID of the `H`, if needed), so for these modes, the numbers are not going to be shorter.

Still, the placement of `a1` is probably not important, and if both you and Manuel think the first scheme is better, let's use it.

So let's discuss it briefly tomorrow one final time and nail down and final scheme (either `d1d2a1b1b2c1` or `a1d1d2b1b2c1`).

**Comment by yipengsun** [comment]
I read your truth-matching doc and it looks very reasonable. I did some formatting on your doc.

For the code, I just use `clang-format` to make sure the style is consistent, then did some minor manual editing.

Don't worry about the style when you try updating that file, because `clang-format` will automatically bring your code to the style-standard set for this project--Code as you wish!
